test/models/bot/tagger_test.rb

```diff
@@ -0,0 +1,117 @@
+require_relative '../../test_helper'
+
+class Bot::TaggerTest < ActiveSupport::TestCase
+  def setup
+    super
+    @TaggerBot = create_tagger_bot(name: "tagger", login: "tagger")
+    @AlegreBot = create_alegre_bot(name: "alegre", login: "alegre")
+    @TaggerBot.approve!
+    @AlegreBot.approve!
+    p = create_project
+    p.team.save!
+    @AlegreBot.install_to!(p.team)
+    @TaggerBot.install_to!(p.team)
+    @team = p.team
+    m = create_claim_media quote: 'I like apples'
+    @pm = create_project_media project: p, media: m
+    create_flag_annotation_type
+    create_extracted_text_annotation_type
+    Sidekiq::Testing.inline!
+    @auto_tag_prefix = "!"
+  end
+
+  def teardown
+    super
+  end
+
+  def create_tagger_bot(_options = {})
+    Bot::Tagger.new(_options)
+  end
+
+  test "should get tag text" do
+    # Method signature
+    # get_tag_text(tag_id,auto_tag_prefix,ignore_autotags)
+    pm1 = create_project_media quote: "Rain", team: @team
+    tag1 = Tag.create!(annotated:pm1, annotator: BotUser.get_user('tagger'), tag: @auto_tag_prefix+"test")
+
+    pm2 = create_project_media quote: "Rain", team: @team
+    tag2 = Tag.create!(annotated:pm2, annotator: BotUser.get_user('tagger'), tag: "test")
+
+    # Get tag text keeping autotags
+    # Should return tag with prefix stripped
+    assert_equal "test", Bot::Tagger.get_tag_text(tag1[:data][:tag], @auto_tag_prefix, false)
+
+    # Get tag text ignoring autotags
+    assert_nil Bot::Tagger.get_tag_text(tag1[:data][:tag], @auto_tag_prefix, true)
+    assert_equal "test", Bot::Tagger.get_tag_text(tag2[:data][:tag], @auto_tag_prefix, true)
+
+    # Get tag text for invalid id
+    # Should return nil
+    assert_nil Bot::Tagger.get_tag_text(-999, @auto_tag_prefix, false)
+  end
+
+  test "should tag item" do
```

Comment:
**praise:** it's so great that you added comprehensive tests around this!
**suggestion** in future I would suggest breaking this into more granular tests so that we get more meaningful failures if it fails midway through, and we can see all that fails without having to rerun. I think the comments you added to mark different "shoulds" would make great test divisions since they imply different expectations for different setups.

Reply:
Yeah, I was trying to balance between 
* more granular tests and 
* not having repeated setup code / not having too many tests that could slow down CI given this is quite a small, periphery bot at least for now
At the time I thought it made sense to group the three cases here that all test one function with the same setup code, but I can see the value of breaking them apart too. Thank you!